### PR TITLE
feat: mongodb protocol

### LINKED ISSUE #12: Complete Wire Protocol
Complete the foundation of the driver: the wire protocol.

### PR DESCRIPTION
**Pull request by lucacasonato**
~WIP implementation of the MongoDB wire protocol. The basic implementation of OP_QUERY and OP_REPLY seem to work. We are now just constrained by not being able to parse the BSON documents that are returned.~

After reading through the docs, it seems we only need `OP_MSG`. The rest are legacy ops that are not anymore needed.

### CODE REVIEW

**Comment by lucacasonato** [comment]
This should be all we need for the protocol layer. Once this is landed we can start on the real driver.

**Review by ebebbington — COMMENTED** [review]
LGTM

File: bson/src/lib.rs [hunk 1]
```diff
@@ -7,21 +7,26 @@ mod encoder;
...
-        .map_err(|err| format!("error serializing document to json: {}", err.to_string()))?)
+    let document = bson::Document::from_reader(&mut x).map_err(|err| {
+        js_sys::Error::new(&format!("error parsing document: {}", err.to_string()))
+    })?;
+    Ok(serde_json::to_string(&document).map_err(|err| {
+        js_sys::Error::new(&format!(
+            "error serializing document to json: {}",
+            err.to_string()
+        ))
+    })?)
```
Comment (notfilippo) [inline bson/src/lib.rs]: returning `JsValue::from_serde(&example)` is a bit more standard than returning a rust String

**Review by notfilippo — CHANGES_REQUESTED** [review]
Small nitpick.

File: bson/src/lib.rs (hunk 1 above)
Comment (lucacasonato) [reply bson/src/lib.rs]: I wonder if that introduces an extra copy. I will investigate.

**Review by notfilippo — APPROVED** [review]
LGTM

**Pull request merged**